runtime/integration-tests/src/generic/cases/block_rewards.rs

```diff
@@ -0,0 +1,199 @@
+use cfg_traits::rewards::{AccountRewards, CurrencyGroupChange, DistributedRewards};
+use cfg_types::tokens::CurrencyId;
+use frame_support::{assert_ok, dispatch::RawOrigin};
+use runtime_common::{
+	apis::{runtime_decl_for_rewards_api::RewardsApiV1, RewardDomain},
+	instances::BlockRewards,
+};
+use sp_runtime::traits::{Get, Zero};
+
+use crate::{
+	generic::{
+		config::Runtime,
+		env::Env,
+		envs::runtime_env::RuntimeEnv,
+		utils,
+		utils::{
+			currency::cfg,
+			genesis::{self, Genesis},
+		},
+	},
+	utils::accounts::Keyring,
+};
+
+crate::test_for_runtimes!(all, block_rewards_api);
+fn block_rewards_api<T: Runtime>() {
+	const STAKER: Keyring = Keyring::Alice;
+
+	RuntimeEnv::<T>::default().parachain_state_mut(|| {
+		let group_id = 1u32;
+		let amount = 100 * CFG;
+
+		utils::give_balance::<T>(STAKER.id(), T::ExistentialDeposit::get() + amount);
+
+		assert_ok!(pallet_rewards::Pallet::<T, BlockRewards>::attach_currency(
+			CurrencyId::Native,
+			group_id,
+		));
+
+		assert_ok!(pallet_rewards::Pallet::<T, BlockRewards>::deposit_stake(
+			CurrencyId::Native,
+			&STAKER.id(),
+			amount,
+		));
+
+		assert_ok!(
+			pallet_rewards::Pallet::<T, BlockRewards>::distribute_reward(200 * CFG, [group_id])
+		);
+
+		assert_eq!(
+			T::Api::list_currencies(RewardDomain::Block, STAKER.id()),
+			vec![CurrencyId::Native]
+		);
+
+		assert_eq!(
+			T::Api::compute_reward(RewardDomain::Block, CurrencyId::Native, STAKER.id()),
+			Some(200 * CFG)
+		)
+	});
+}
+
+crate::test_for_runtimes!([development], collator_list_synchronized);
+fn collator_list_synchronized<T: Runtime>() {
+	RuntimeEnv::<T>::from_parachain_storage(
+		Genesis::default()
+			.add(genesis::balances::<T>(
+				T::ExistentialDeposit::get() + cfg(100),
+			))
+			.add(genesis::invulnerables::<T>(vec![Keyring::Admin]))
+			.add(genesis::session_keys::<T>())
+			.add(genesis::block_rewards::<T>(vec![Keyring::Admin]))
+			.storage(),
+	)
+	.parachain_state_mut(|| {
+		let collators_1 = vec![Keyring::Alice.id(), Keyring::Bob.id()];
+		let collators_2 = vec![
+			Keyring::Charlie.id(),
+			Keyring::Dave.id(),
+			Keyring::Eve.id(),
+			Keyring::Ferdie.id(),
+		];
+
+		// SESSION 0 -> 1;
+		apply_and_check_session::<T>(1, collators_1.clone(), vec![]);
+
+		// SESSION 1 -> 2;
+		apply_and_check_session::<T>(3, collators_2.clone(), vec![]);
+
+		// SESSION 2 -> 3;
+		apply_and_check_session::<T>(7, vec![], vec![Keyring::Alice.id()]);
+
+		assert!(collators_1.iter().all(|c| has_reward::<T>(c)));
+		assert!(collators_2.iter().all(|c| !has_reward::<T>(c)));
+
+		// SESSION 3 -> 4;
+		apply_and_check_session::<T>(6, vec![], vec![]);
+
+		assert!(collators_2.iter().all(|c| has_reward::<T>(c)));
+	});
+}
+
+fn apply_and_check_session<T: Runtime>(
+	collator_count: u32,
+	joining: Vec<AccountId>,
+	leaving: Vec<AccountId>,
+) {
+	for collator in &joining {
+		pallet_collator_selection::Pallet::<T>::register_as_candidate(
+			RawOrigin::Signed(collator.clone()).into(),
+		)
+		.unwrap();
```

Comment:
I can't call this in `altair` and `centrifuge` because I get a `ValidatorNotRegistered` error. 

@wischli, do you have any idea why this happens?

Reply:
In terms of `pallet_collator_selection::Config`, Dev and Altair/Centrifuge differ in the `ValidatorRegistration` trait implementor: Dev uses `Sessions` whereas production uses the `CollatorAllowlist`. Therefore, dev only requires session keys to be set for those addresses, which are registered. Production however requires the accounts to be whitelisted via the `pallet_collator_allowlist` beforehand. This can done via the `add` extrinsic of that pallet. This also requires the address to have their session key set.

Reply:
Thanks so much for this answer @wischli, it helps me a lot!

Reply:
@wischli, do you think we should align this for all runtimes? Why we are not using collator allowlist for dev?

cc @mustermeiszer

Reply:
Functionally, it does not make sense to me to add an extra gate to collators for our dev runtime IMO. I suppose the only achievement for us would be that this test works for all runtimes? 

I definitely dont want to block your proposal, just questioning the effort.

Reply:
Nono, it's not needed for the test; we can discriminate the behavior per runtime (at least it seems so by now). I asked because `development` is our way of proving stuff that later will go to `centrifuge`. So maybe it deserves that the demo also uses collator allowlist. Just a thought!